components/script/dom/bindings/str.rs

```diff
@@ -208,6 +208,95 @@ impl DOMString {
         self.0.truncate(last_non_whitespace);
         let _ = self.0.splice(0..first_non_whitespace, "");
     }
+
+    /// Valid date string should be YYYY-MM-DD, YYYY must be four or more
+    /// https://html.spec.whatwg.org/multipage/#valid-date-string
+    pub fn is_valid_date_string(&mut self) -> bool {
+        // Step 2
+        let value = &*(self.0);
+        let value_vec: Vec<&str> = value.split('-').collect();
+        if !(value_vec.len() == 3) {
+            return false;
+        }
+        let year = value_vec[0];
+        let month = value_vec[1];
+        let day = value_vec[2];
+
+        // Step 1
+        let month_string = format!("{}-{}", year, month);
```

Comment:
This is still not OK. `format!` heap-allocates a String, and I believe we don't need any heap allocation at all.